src/v/archival/archiver_manager.cc

```diff
@@ -0,0 +1,1051 @@
+    retry_chain_node _rtc;
+    retry_chain_logger _ctxlog;
+
+    // archiver (only if is_leader)
+    // note: this will be changed to ntp_archiver_api in the future
+    ss::lw_shared_ptr<ntp_archiver> _archiver;
+    // currently active background operation (start/stop archiver)
+    std::optional<ss::future<>> _bg_operation;
+    // barrier is used by stop method
+    std::optional<ss::promise<>> _barrier;
+};
+
+struct managed_partition : public managed_partition_fsm::state_machine_t {
+    managed_partition(
+      model::ntp ntp,
+      model::node_id broker_id,
+      ss::lw_shared_ptr<cluster::partition> part,
+      ss::lw_shared_ptr<const archival::configuration> config,
+      cloud_storage::remote& remote,
+      cloud_storage::cache& cache,
+      archival::upload_housekeeping_service& housekeeping,
+      ss::gate::holder gh)
+      : managed_partition_fsm::state_machine_t(
+        ntp,
+        broker_id,
+        std::move(part),
+        std::move(config),
+        remote,
+        cache,
+        housekeeping)
+      , _ntp(ntp)
+      , _node_id(broker_id)
+      , _gate_holder(std::move(gh)) {}
+
+    ~managed_partition() {
+        auto st = current_state();
+        vassert(
+          st == managed_partition_state_t::passive,
+          "Unexpected managed_partition state {}",
+          st);
+    }
+
+    /// Acquired leadership notification
+    void notify_leadership_acquired(model::term_id term) {
+        process_event(ev_leadership_acquired{.term = term});
+    }
+
+    /// Lost leadership notification
+    void notify_leadership_lost() { process_event(ev_leadership_lost{}); }
+
+    /// Stop managed partition instance.
+    /// The object can't be safely used when returned future becomes ready.
+    ss::future<> stop() {
+        // The shutdown is relying on the following mechanism:
+        // - We can guarantee that the archiver is stopped by transitioning
+        //   into the _passive_ state.
+        // - If the FSM is already in this state the exit.
+        // - Otherwise we need to wait for the transition to complete, in order
+        //   to do this we:
+        //   - Installing 'promise' object into the _passive_ state.
+        //   - Triggering 'shutdown' event in FSM.
+        //   - Awaiting the future produced from the promise object we just
+        //     installed.
+        //   - When the _passive_ state is entered it first checks if the
+        //     promise is installed, if this is the case it sets the promise.
+        //   - If the _passive_ state was entered after exception was thrown
+        //     the 'set_exception' method of the promise will be set to
+        //     propagate the error.
+        if (current_state() == managed_partition_state_t::passive) {
+            co_return;
+        }
+        auto& st = get_state<st_passive&>();
+        st.trigger = ss::promise<>();
+        process_event(ev_shutdown{});
+        co_await st.trigger->get_future();
```

Comment:
Under the hood, is this saying, "next time the stm is in `st_passive`, it will have this promise"? I think that makes sense, but I'm wondering what it means when `ev_shutdown` defers (assuming `defer` means "no transition). Is it possible we are running `stop()` after we've already unregistered the notifications (looking at `archiver_manager::stop()`)

Reply:
The `ev_shutdown` is simply delayed until the next state. So if the FSM is in the `st_starting_async` and `ev_shutdown` is triggered it won't do anything until the startup will finish and the FSM will go into the `st_active` state. This state reacts to the event and this will trigger transition to the `st_stopping_async` state and eventually to the `st_passive`. This will lead to the promise being set so the future will become ready.